### ISSUE TITLE
Expose preference key to Converter

### ISSUE THREAD

**Issue by onomated** [f2prateek/rx-preferences#100]
The change from `Adapter` to `Converter` interface (#75) breaks some functionality use cases where serializing/deserializing from shared preferences depends on the preference key.
For example, I have an EncryptedPreferenceAdapter that worked as follows with the previous Adapter interface:
```java
    @Override
    public Optional<T> get(@NonNull String key, @NonNull SharedPreferences preferences) {
            final String serialized = preferences.getString(key, null);
            if (serialized == null)
                return Optional.empty();
            final String decrypted = Obfuscator.decryptFromStorage(key, serialized);
            return Optional.ofNullable(decrypted);
    }

    @Override
    public void set(@NonNull String key, @NonNull Optional<T> optionalValue, @NonNull SharedPreferences.Editor editor) {
            if (optionalValue.isPresent()) {
                editor.putString(key, Obfuscator.encryptForStorage(key, optionalValue.get()));
            }
            else {
                editor.putString(key, null);
            }
    }
```
Can Converters be updated to provide the preference key?  Or can adapters be resurrected since there may be apps out in the wild that have adapters serializing to non-string types?

**Comment by f2prateek** [comment]
yes, this was a known limitation of switching over. I'll leave this issue open for others to chime in so we can get a sense of how common  this request is.

**Comment by onomated** [comment]
Well, this handicaps my ability to update to future versions of this component for this specific app at least.  From @NightlyNexus comment in #75 sounds like this change wasn't made to address fix or address any issues, it was made purely from an interface design point of view.  Also, my particular scenario was not one of the disadvantages considered in that conversation.  Am I missing something here? Or is there a reason why the preference key is now hidden from clients?

**Comment by f2prateek** [comment]
Jake mentioned it https://github.com/f2prateek/rx-preferences/pull/75#issuecomment-287572822

**Comment by onomated** [comment]
That's a different consideration in that he's talking about the data-type of the serialized data, and I'm talking about access to the preference key so I can take action based on the key value.  With regards to Jake's comment, he does state that it is a **disadvantage**.  Ultimately, there is less flexibility to developers with no perceived gain otherwise.  The change in the interface was not solving a problem.

**Comment by f2prateek** [comment]
Got it. Generally yes, we knew it was going to limit flexibility.

The gain is that there's smaller API surface exposed and less chances of developers making mistakes.

**Comment by m4xp1** [comment]
I think the converter interface is very good. And your case is quite specific. Because I think that for you it will be a good option that something like this: 
```
  public class KeyConverter implements Preference.Converter<String> {

        private final String key;

        public KeyConverter(String key) {
            this.key = key;
        }

        @Override
        public @NonNull String deserialize(@NonNull String serialized) {
            return Obfuscator.decryptFromStorage(key, serialized);
        }

        @Override
        public @NonNull String serialize(@NonNull String value) {
            return Obfuscator.encryptForStorage(key, value);
        }
    }
```
What say?

**Comment by PaulWoitaschek** [comment]
I think it's actually a good change to hide the key.

For example I had an implementation where I did not actually use the key in my adapter but used it to prefix mutliple values and write them to the shared preferences.

So I did like:
```kotlin
editor
  .putString("$key_date", user.date.toString()
  .putString("$key_name", user.name)
```

However that caused issues because `Preference.get()` was checking for the key. And that key was never used.

So this change eliminates this error as it abstracts the whole thing one step further.
